scylla/src/transport/cluster.rs

```diff
@@ -476,6 +458,21 @@ impl ClusterData {
     pub fn replica_locator(&self) -> &ReplicaLocator {
         &self.locator
     }
+
+    /// Returns iterator of connections to all shards.
+    ///
+    /// By nonemptiness invariant of `self.known_peers`, the returned iterator is nonempty.
+    pub(crate) fn iter_working_connections(
+        &self,
+    ) -> impl Iterator<Item = Result<Arc<Connection>, QueryError>> + '_ {
+        let peers = &self.known_peers;
+        peers
+            .values()
+            .map(|node| node.get_working_connections())
+            .flatten_ok()
+        // By an invariant `self.known_peers` is nonempty, so the returned iterator
+        // is nonempty, too.
```

Comment:
Why is that? If all node pools are in `MaybePoolConnections::Broken` state, every call to `node.get_working_connections()` will fail.

Reply:
That's true, and hence the resulting iterator will consist of only `Err` variants.
`flatten_ok()` does not remove any `Err` variants; what it does is flattening the `Ok` variants; that is, `Ok(Ok(x))` becomes `Ok(x)`.